p2p/src/net/mock/transport/impls/stream_adapter/wrapped_transport/utils.rs

```diff
@@ -36,7 +36,7 @@ impl<'a, S: StreamAdapter<T::Stream>, T: TransportSocket> Future for HandshakeFu
             for i in 0..self.0.len() {
                 match Future::poll(self.0[i].0.as_mut(), cx) {
                     Poll::Ready(res) => {
-                        let (_, addr) = self.0.remove(i);
+                        let (_, addr) = self.0.swap_remove(i);
```

Comment:
First, really impressive that rust provides this. I'm very happy.

Second, your usage of the implementation is now incorrect. This is the complexity I was talking about. It would've been hell to implement this whole thing ourselves, but while rust removed part of the burden, this still has corner cases that should be treated with care. The problem with your current implementation is that we should subtract `1` from `i` to re-account for the last element that replaced element number `i`. A corner case there would be when `i == 0`, which will cause an integer underflow. Funnily enough, if you blindly subtract `1` in C++, it would work, because the next `+1` in the loop will cause another overflow and fix itself. But now in rust, with `usize`, this should be treated with care because it'll get asserts to fire in debug mode. Please isolate this treatment in a function that can be tested and don't be hasty in fixing this.

Reply:
Yes, you are absolutely right about such corner cases in general. But here when a future is complete and removed we either return result with `Poll::Ready` (if handshake succeeds) or start looping over (if handshake fails). So we don't need to bother adjusting the index.

Reply:
Alrighty. Please add a comment above the remove saying "Index `i` will be skipped after the swap, but this is OK because the next poll will go over it again".

Reply:
Feel free to resolve this comment after adding the comment in the code.

Reply:
Outer loop was removed and we now always return.